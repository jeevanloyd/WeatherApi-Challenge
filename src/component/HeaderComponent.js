import React from 'react'

// This is just an Header Component 
const HeaderComponent = () => {
  return (  
    <div className='d-flex bg-dark text-white justify-content-center py-4'>
        Simple Weather App
    </div>
  )
}

export default HeaderComponent;