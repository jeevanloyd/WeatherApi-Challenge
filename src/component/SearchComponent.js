import React, { useState } from "react";
import { API_KEY, BASE_SEARCH_URL } from "../ApiHelpers";
const axios = require("axios").default;

// Search Component : User can search for specfic location . 
// This method returns the geolocation cordinates ie.  Lat and lon
// There are few drawbacks with this Api, we might need to pay to upgrade and get more benifits from this Api.
//This Api is restricted to max 5 cities and for this reason, i have restricted this search only for Australia.

const SearchComponent = (props) => {

    const [cityName, setcityName] = useState("");
    const onSearchBtnClick = () => {

        if (cityName !== "") {
            props.setisSearchActive(true)
            var SEARCH_API_URL =
                // eslint-disable-next-line no-useless-concat
                BASE_SEARCH_URL + "=" + cityName + "&limit=1&appid=" + API_KEY;
            axios
                .get(SEARCH_API_URL)
                .then(function (response) {

                    var temp_coordinate = {
                        lat: response["data"][0].lat,
                        lon: response["data"][0].lon
                    }
                    props.setGeoLocation(temp_coordinate)

                })
                .catch(function (error) {
                    alert("Please select a correct location in Australia.")
                    console.log(error);
                });
        }
    };

    const onClearBtnClick = () => {
        if (cityName !== '') {
            document.getElementById("textbox").value = "";
            setcityName('')
        }
    }

    const SearchtextBox = (value) => {
        if (value !== '')
            setcityName(value)

    }

    return (
        <div className="row">
            <div className="mb-3 row justify-content-end">

                <input
                    type="text"
                    id="textbox"
                    class="form-control"
                    placeholder="Search City"
                    aria-label="search City"
                    aria-describedby="basic-addon1"
                    data-toggle="tooltip" data-placement="top" title="Please enter CityName,CountryCode(Ex:geelong,AU)"
                    value={cityName}
                    onChange={(e) => SearchtextBox(e.target.value)}
                />
                <span className=" justify-content-end p-2 align-bottom">
                    <button
                        className="btn btn-primary"
                        style={{ height: "fit-content" }}
                        onClick={onSearchBtnClick}
                    >
                        Find
                    </button>
                    <span> <button
                        className="btn btn-primary
                        "
                        style={{ height: "fit-content" }}
                        onClick={onClearBtnClick}
                    >
                        Clear
                    </button></span>
                    
                </span>
            </div>
        </div>
    );
};

export default SearchComponent;
