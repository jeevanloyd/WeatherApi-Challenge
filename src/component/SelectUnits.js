import React from "react";
// This component is specailly designed for change units. User can set to choose between Imperial vs Metric

export const SelectUnits = (props) => {
    return (
        <div className="row">
            <div class="input-group mb-3">
                <div class="input-group-prepend">
                    <label class="input-group-text" for="unitsselect">
                        Choose Units
                    </label>
                </div>
                <select
                    class="custom-select"
                    id="unitsselect"
                    value={props.metric}
                    onChange={(e) => props.setMetric(e.target.value)}
                >
                    <option value="metric" selected>
                        Metric
                    </option>
                    <option value="imperial">Imperial</option>
                </select>
            </div>
        </div>
    );
};
