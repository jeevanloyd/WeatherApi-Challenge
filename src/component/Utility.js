import moment from "moment";

// general Util file 

export const convertDate = (milliSecond) => {
    return moment
        .unix(milliSecond)
        .locale("zh-tw")
        .format("dddd, MMMM Do YYYY");
};

export const getDayName = (milliSecond) => {
    return moment
        .unix(milliSecond)
        .locale("zh-tw")
        .format('dddd');
};