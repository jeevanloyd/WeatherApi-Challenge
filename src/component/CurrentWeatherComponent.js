import React from "react";
import { convertDate } from "./Utility";

// This components is reused , It gives data for current weather based on device location and based on geocordinates from seach API
const CurrentWeatherComponent = (props) => {
    let weatherData = props.weather;
    return (
        <div className="row justify-content-between">
            <div className="col-12 col-md-4 bg-white rounded border">
                <div className="row">
                    <h3>{weatherData.name}</h3>
                </div>
                <div className="row justify-content-center">{convertDate(weatherData.dt)}</div>
                <div className="row justify-content-center">{weatherData.weather[0].main}</div>
                <div className="row d-flex align-top justify-content-center ">
                    <img
                        src={`icons/${weatherData.weather[0].icon}.png`}
                        className="icon-small"
                        alt="weather"
                        width={20}
                    />
                    
                </div><div className="row justify-content-center ">
                            <span><h3><span className="font-weight-bold">{ Math.round(weatherData.main.temp)}</span><sup>{props.metric ==='metric'?'C' : 'F' }</sup></h3></span>
                        </div>
            </div>
            <div className="col-12 col-md-4 justify-content-end ">
                <div className="row w-100"><span className="text-muted">feels like: &nbsp; </span><span className="font-weight-bold">{Math.round(weatherData.main.feels_like)}<sup>{props.metric ==='metric'?'C' : 'F' }</sup> </span></div>
                <div className="row w-100"><span className="text-muted">Humidity:&nbsp; </span><span className="font-weight-bold">{weatherData.main.humidity} %</span></div>
                <div className="row w-100"><span className="text-muted">Wind Speed: &nbsp;</span><span className="font-weight-bold">{weatherData.wind.speed} {props.metric ==='metric'?'m/s' : 'mph' }</span></div>
            </div>
        </div>
    );
};

export default CurrentWeatherComponent;
