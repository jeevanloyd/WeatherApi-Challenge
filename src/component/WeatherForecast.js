import React from "react";
import { getDayName } from "./Utility";
 // Drawbacks of API,
// The weather forecast data ( freely available ) , has data for 5 days

const WeatherForecast = (props) => {
    let weatherForeCastData = props.forecastweather;
    let foreCastArr = [];
    for (let i = 0; i < weatherForeCastData.list.length; i += 8) {
        foreCastArr.push(weatherForeCastData.list[i]);
    }
    return (
        <div className="row justify-content-between w-100 mt-4">
            {foreCastArr.map((item, index) => {
                return (
                    <div className="col-12 col-md-2 bg-white rounded border" key={index}>
                        <div className="row justify-content-center py-2">
                            <h6>{index === 0 ? "Today" : getDayName(item.dt)}</h6>
                        </div>
                        <div className="row justify-content-center">
                            <img
                                src={`icons/${item.weather[0].icon}.png`}
                                className="icon-small"
                                alt="weather"
                                width={40}
                            />
                        </div>
                        <div className="row justify-content-center">
                            <span><span className="font-weight-bold">{ Math.round(item.main.temp)}</span><sup>{props.metric ==='metric'?'C' : 'F' }</sup></span>
                        </div>
                        <div className="row justify-content-center">
                            { item.weather[0].main}
                        </div>
                    </div>
                );
            })}
        </div>
    );
};

export default WeatherForecast;
