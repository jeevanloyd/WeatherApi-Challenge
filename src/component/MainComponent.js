import React, { useEffect, useState } from "react";
import { BASE_URL, API_KEY, FORECAST_URL } from "../ApiHelpers";
import SearchComponent from "./SearchComponent";
import { SelectUnits } from "./SelectUnits";
import CurrentWeatherComponent from "./CurrentWeatherComponent";
import WeatherForecast from "./WeatherForecast";

const axios = require("axios").default;
// This is the main component
// We will use 2 API to fetch CurrentWeather (Weather details shown onload of application) and forecast whether for upcoming 5 days.
// This application is done with React , Javascript, HTML, Boots

const MainComponent = () => {
    const [todayWeather, setTodayWeather] = useState({});
    const [isWeatherFetch, setisWeatherFetch] = useState(false);
    const [forecastWeather, setForecastWeather] = useState({});
    const [isForecastFetch, setisForecastFetch] = useState(false);
    const [metric, setMetric] = useState("metric");
    const [isSearchActive, setisSearchActive] = useState(false);
    const [geoLocation, setGeoLocation] = useState({});

    const getCurrentWeather = (api_url) => {
        axios
            .get(api_url)
            .then(function (response) {
                setisWeatherFetch(true);
                setTodayWeather(response.data);
            })
            .catch(function (error) {
                console.log(error);
            });
    };

    const getForecastWeather = (api_url) => {
        axios
            .get(api_url)
            .then(function (response) {
                setisForecastFetch(true);
                setForecastWeather(response.data);
            })
            .catch(function (error) {
                console.log(error);
            });
    };
// we will show current weather by default , unless user has searched for location.
// For user search [isSearchActive is set to true] , we must pass cordinates/geolocation , in order to get the weather data.
    useEffect(() => {
        if (isSearchActive) {
            const API_URL =
                BASE_URL +
                `lat=${geoLocation.lat}&lon=${geoLocation.lon}&units=${metric}&appid=` +
                API_KEY;
            const FORECAST_API =
                FORECAST_URL +
                `lat=${geoLocation.lat}&lon=${geoLocation.lon}&units=${metric}&appid=` +
                API_KEY;

            getCurrentWeather(API_URL);
            getForecastWeather(FORECAST_API);
        } else {
            // The Geolocation.getCurrentPosition() method is used to get the current position of the device.

            navigator.geolocation.getCurrentPosition((position) => {
                const API_URL =
                    BASE_URL +
                    `lat=${position.coords.latitude}&lon=${position.coords.longitude}&units=${metric}&appid=` +
                    API_KEY;

                const FORECAST_API =
                    FORECAST_URL +
                    `lat=${position.coords.latitude}&lon=${position.coords.longitude}&units=${metric}&appid=` +
                    API_KEY;

                getCurrentWeather(API_URL);
                getForecastWeather(FORECAST_API);
            });
        }
    }, [metric, geoLocation,isSearchActive]); // It will call the api , once these values are changed.
    return (
        <div className="container-fluid mt-3 bg-light py-5 px-5">
            <div className="row mb-3 justify-content-between">
                <div className="col-md-4">
                    <SelectUnits metric={metric} setMetric={setMetric} />
                </div>
                <div className="col-md-6 d-flex">
                    <SearchComponent
                        isSearchActive={isSearchActive}
                        setisSearchActive={setisSearchActive}
                        setGeoLocation={setGeoLocation}
                    />
                </div>
            </div>

            {isWeatherFetch ? (
                <CurrentWeatherComponent weather={todayWeather} metric={metric} />
            ) : (
                <div class="text-center">
                    <div class="spinner-border" role="status">
                        <span class="visually-hidden">Loading...</span>
                    </div>
                </div>
            )}

            <div className="row">
                {isForecastFetch ? (
                    <WeatherForecast
                        forecastweather={forecastWeather}
                        metric={metric}
                    />
                ) : (
                    <div>  Loading...
                    </div>
                )}
            </div>

        </div>
    );
};

export default MainComponent;
