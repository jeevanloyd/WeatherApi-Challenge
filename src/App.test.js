
import renderer from 'react-test-renderer';
import SearchComponent from './component/SearchComponent';
import HeaderComponent from './component/HeaderComponent';
import { cleanup } from '@testing-library/react';
import CurrentWeatherComponent from './component/CurrentWeatherComponent';
import { SelectUnits } from './component/SelectUnits';

afterEach(cleanup);

describe('Render Search Component', () => {

  it('snapshot testing for Search component', () => {
    const component = renderer.create(
      <SearchComponent />,
    );
    let Searchtree = component.toJSON();
    expect(Searchtree).toMatchSnapshot();
  })
})

describe('renders header component', () => {

  it('snapshot testing for Header component', () => {
    const component1 = renderer.create(
      <HeaderComponent />,
    );
    let Header = component1.toJSON();
    expect(Header).toMatchSnapshot();
  })
})

describe.skip('renders Todays weather component', () => {
  let weatherdata = {

    weather: [
      {
        id: 803,
        main: "Clouds",
        description: "broken clouds",
        icon: "04d"
      }
    ],
    main: {
      temp: 14.6,
      "feels_like": 13.87,
      "humidity": 67
    },

    wind: {
      speed: 4.92,
    },

    dt: 1660791659,
    name: "Richmond",

  }

  it('snapshot testing for Header component', () => {
    
    const component3 = renderer.create(
      <CurrentWeatherComponent weatherData={weatherdata} />,
    );

    let Header1 = component3.toJSON();
    expect(Header1).toMatchSnapshot();
  })
})

describe('renders select Units component', () => {

  it('snapshot testing for SelectUnits component', () => {
    const component2 = renderer.create(
      <SelectUnits />,
    );
    let unitsTree = component2.toJSON();
    expect(unitsTree).toMatchSnapshot();
  })
})



