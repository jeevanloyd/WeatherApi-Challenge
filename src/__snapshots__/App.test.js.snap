// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Search Component  snapshot testing for Search component 1`] = `
<div
  className="row"
>
  <div
    className="mb-3 row justify-content-end"
  >
    <input
      aria-describedby="basic-addon1"
      aria-label="search City"
      class="form-control "
      id="textbox"
      onChange={[Function]}
      placeholder="Search City"
      type="text"
      value=""
    />
    <span
      className="  justify-content-end p-2 align-bottom"
    >
      <button
        className="btn btn-primary "
        onClick={[Function]}
        style={
          Object {
            "height": "fit-content",
          }
        }
      >
        Find
      </button>
      <span>
         
        <button
          className="btn btn-primary "
          onClick={[Function]}
          style={
            Object {
              "height": "fit-content",
            }
          }
        >
          Clear
        </button>
      </span>
    </span>
  </div>
</div>
`;

exports[`renders header component  snapshot testing for Header component 1`] = `
<div
  className="d-flex bg-dark text-white justify-content-center py-4"
>
  Simple Weather App
</div>
`;

exports[`renders select Units component  snapshot testing for SelectUnits component 1`] = `
<div
  className="row"
>
  <div
    class="input-group mb-3"
  >
    <div
      class="input-group-prepend"
    >
      <label
        class="input-group-text"
        for="unitsselect"
      >
        Choose Units
      </label>
    </div>
    <select
      class="custom-select"
      id="unitsselect"
      onChange={[Function]}
    >
      <option
        selected={true}
        value="metric"
      >
        Metric
      </option>
      <option
        value="imperial"
      >
        Imperial
      </option>
    </select>
  </div>
</div>
`;
